Selection does not update during autoscroll when selecting with gesture or mouse.

megangardner opened the pull request:
#### 35b1fa78171adc4d49d120994b5bebe9ee48b0d4
<pre>
Selection does not update during autoscroll when selecting with gesture or mouse.
<a href="https://bugs.webkit.org/show_bug.cgi?id=287600">https://bugs.webkit.org/show_bug.cgi?id=287600</a>
<a href="https://rdar.apple.com/144744443">rdar://144744443</a>

Reviewed by Wenson Hsieh.

Similar to <a href="https://bugs.webkit.org/show_bug.cgi?id=287454">https://bugs.webkit.org/show_bug.cgi?id=287454</a>

When a user selects text using the mouse or the tap-and-a-half gesture,
and scrolls far enough to start an autoscroll, they don&apos;t generate
additional touches unless their finger moves. Thus if you
hold your finger still while autoscrolling, the selection doesn&apos;t update.
In order to for a selection to update, we keep track of the position of
the last touch and use it to update the selection when the view is scrolled,
and only update based on the gesture or touch that is currently active. We
only allow updates while autoscroll is active.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView cancelAutoscroll]):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView updateSelection]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:hasFocusedElement:respectSelectionAnchor:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPointAndBoundary:textGranularity:textInteractionSource:completionHandler:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::startAutoscrollAtPosition):
(WebKit::WebPageProxy::cancelAutoscroll):

Canonical link: <a href="https://commits.webkit.org/290367@main">https://commits.webkit.org/290367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec326888197d61ac245cd8be04d968a636519248

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89801 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/9330 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44692 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94798 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40574 "Built successfully") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/130/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17608 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/69158 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/26780 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92802 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/130/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/81480 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/49519 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/130/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39707 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/130/builds/9717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96624 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16988 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/12472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/78030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/17244 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/77294 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/77353 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19097 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21794 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/20371 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/10165 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16999 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/22320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16740 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/20192 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18523 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
```diff
@@ -445,14 +445,35 @@ struct ImageAnalysisContextMenuActionData {
+        ExtentPointAndBoundary,
+    };
+
+    SelectionInteractionType _isInSelectionInteraction;
```
Maybe this should be renamed to `_selectionInteractionType` for consistency?

megangardner replied:
Good call

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5265,6 +5270,8 @@ - (void)changeSelectionWithGestureAt:(CGPoint)point withGesture:(WKBEGestureType
     logTextInteraction(__PRETTY_FUNCTION__, self.textInteractionLoupeGestureRecognizer, point, state, gestureType, std::nullopt, flags);
 
+    ALWAYS_LOG_WITH_STREAM(stream << "<MMG> changeSelectionWithGestureAt");
+
```
Spurious logging.
```suggestion
```

megangardner replied:
This and the below log are somehow not in my local branch, I'm not sure what happened.

whsieh commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -5663,6 +5677,9 @@ - (void)updateSelectionWithExtentPoint:(CGPoint)point completionHandler:(void (^
 - (void)updateSelectionWithExtentPoint:(CGPoint)point withBoundary:(UITextGranularity)granularity completionHandler:(void (^)(BOOL selectionEndIsMoving))completionHandler
 {
+    ALWAYS_LOG_WITH_STREAM(stream << "<MMG> updateSelectionWithExtentPoint ");
+    WTFReportBacktraceWithPrefix("<MMG>");
```
More logging

```suggestion
```

megangardner replied:
I swear I pulled these out... :(

megangardner committed:
Selection does not update during autoscroll when selecting with gesture or mouse.
https://bugs.webkit.org/show_bug.cgi?id=287600
rdar://144744443

Reviewed by Wenson Hsieh.

Similar to https://bugs.webkit.org/show_bug.cgi?id=287454

When a user selects text using the mouse or the tap-and-a-half gesture,
and scrolls far enough to start an autoscroll, they don't generate
additional touches unless their finger moves. Thus if you
hold your finger still while autoscrolling, the selection doesn't update.
In order to for a selection to update, we keep track of the position of
the last touch and use it to update the selection when the view is scrolled,
and only update based on the gesture or touch that is currently active. We
only allow updates while autoscroll is active.

* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView setUpInteraction]):
(-[WKContentView cleanUpInteraction]):
(-[WKContentView cancelAutoscroll]):
(-[WKContentView changeSelectionWithGestureAt:withGesture:withState:withFlags:]):
(-[WKContentView updateSelection]):
(-[WKContentView changeSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithTouchAt:withSelectionTouch:baseIsStart:withFlags:]):
(-[WKContentView updateSelectionWithExtentPoint:withBoundary:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPoint:hasFocusedElement:respectSelectionAnchor:completionHandler:]):
(-[WKContentView updateSelectionWithExtentPointAndBoundary:textGranularity:textInteractionSource:completionHandler:]):
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::startAutoscrollAtPosition):
(WebKit::WebPageProxy::cancelAutoscroll):

Canonical link: https://commits.webkit.org/290367@main
